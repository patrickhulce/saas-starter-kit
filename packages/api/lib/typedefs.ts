import 'express' // tslint:disable-line

declare module 'express-serve-static-core' {
  /* tslint:disable */
  export interface Response {
    body?: any
  }
  /* tslint:enable */
}

export {}
