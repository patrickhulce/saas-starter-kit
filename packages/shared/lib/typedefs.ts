import {ISQLOptions} from 'klay'

export enum ModelID {
  Account = 'account',
  User = 'user',
}

export interface IAccount {
  id?: number
  name: string
  plan: AccountPlan
  createdAt?: Date
  updatedAt?: Date
}

export interface IUser {
  id?: number
  accountId: number
  role: AuthRole
  email: string
  password: string
  firstName: string
  lastName: string
  isVerified: boolean
  verificationKey?: string
  createdAt?: Date
  updatedAt?: Date
}

export enum AuthRole {
  Root = 'root',
  Admin = 'admin',
  User = 'user',
  Anonymous = 'anonymous',
}

export enum Permission {
  RootAccess = '_root_',
  AccountManage = 'account:manage',
  AccountView = 'account:view',
  UserPassword = 'user:password',
  UserProfile = 'user:profile',
  UserView = 'user:view',
}

export enum AccountPlan {
  Free = 'free',
  Bronze = 'bronze',
  Silver = 'silver',
  Gold = 'gold',
}

export interface IConf {
  debug: boolean
  production: boolean
  isUnderTest: boolean

  origin: string
  cdnAppURL: string
  termsOfServiceURL: string
  displayName: string

  secret: string
  database: ISQLOptions

  mailslurp: {apiKey: string}
  sparkpost: {fromAddress: string; apiKey: string; sendToSink: boolean}
}
