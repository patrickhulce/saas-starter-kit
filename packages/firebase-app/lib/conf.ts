import * as functions from 'firebase-functions'
import {IConf} from '../../shared/lib/typedefs'
import conf from '../../shared/lib/conf'

conf.production = !!functions.config().production
conf.apiPathPrefix = '/api'
conf.database.connectionURL = functions.config().sqlURL

export default conf
