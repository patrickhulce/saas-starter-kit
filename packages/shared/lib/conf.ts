import {IConf, JsEnvironment} from './typedefs'

const conf: IConf = {
  debug: true,
  production: false,
  isUnderTest: typeof it === 'function', // tslint:disable-line

  domain: 'THE_DOMAIN',
  displayName: 'THE_PRODUCT_DISPLAY_NAME',
  apiPathPrefix: '/api',

  jsEnvironment: JsEnvironment.Server,
  secret: process.env.SECRET || 'super-secret-string',
  database: {
    connectionURL: process.env.APP_MYSQL_URL || 'mysql://root@localhost/the_product',
  },

  mailslurp: {
    apiKey: process.env.MAILSLURP_API_KEY || '',
  },

  sparkpost: {
    sendToSink: true,
    apiKey: process.env.SPARKPOST_API_KEY || '',
    fromAddress: process.env.SPARKPOST_FROM || '',
  },
}

if (process.env.APP_ENV === 'development') {
  conf.sparkpost.sendToSink = false
}

if (process.env.APP_ENV === 'firebase-development') {
  conf.sparkpost.sendToSink = false
}

if (process.env.APP_ENV === 'firebase-production') {
  conf.debug = false
  conf.production = true
  conf.sparkpost.sendToSink = false
}

if (!conf.sparkpost.fromAddress || conf.isUnderTest) {
  conf.sparkpost.fromAddress = `noreply@${conf.domain}`
}

// tslint:disable-next-line
export default conf
