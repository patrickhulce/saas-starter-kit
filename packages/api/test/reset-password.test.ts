import conf from '../../shared/lib/conf'

import {IState} from './typedefs'

describe('reset password', () => {
  const state: IState = {}
  let passwordResetKey: string

  beforeAll(() => {
    conf.sparkpost.sendToTestMailboxOnly = false
  })

  afterAll(() => {
    conf.sparkpost.sendToTestMailboxOnly = true
  })

  require('./steps/initialize.test')(state)
  require('./steps/account-setup.test')(state)

  describe('reset password', () => {
    it('should reject unauthenticated password changes', async () => {
      const response = await fetch(`${state.apiURL}/v1/users/password-reset`, {
        method: 'PUT',
        body: JSON.stringify({key: 'fake!', password: 'new-password1'}),
        headers: {'content-type': 'application/json'},
      })

      expect(await response.json()).toMatchSnapshot()
      expect(response.status).toBe(400)
    })

    it('should request a password reset', async () => {
      const response = await fetch(`${state.apiURL}/v1/users/password-reset`, {
        method: 'POST',
        body: JSON.stringify({email: state.user.email}),
        headers: {'content-type': 'application/json'},
      })

      expect(response.status).toBe(204)
    })

    it('should send a password reset email', async () => {
      expect(__sparkpostSend).toHaveBeenCalled()
      const payload = __sparkpostSend.mock.calls[1][0]
      const address = payload.recipients[0].address
      passwordResetKey = payload.content.text.match(/key=([\w-]+)/)[1]
      payload.content.text = payload.content.text.replace(/key=[\w-]+/g, 'key=<key>')
      payload.content.html = payload.content.html.replace(/key=[\w-]+/g, 'key=<key>')
      address.email = address.email.replace(/.*@/, '<uuid>@')
      expect(payload).toMatchSnapshot()
    })

    it('should accept password changes', async () => {
      const response = await fetch(`${state.apiURL}/v1/users/password-reset`, {
        method: 'PUT',
        body: JSON.stringify({key: passwordResetKey, password: 'new-password1'}),
        headers: {
          'content-type': 'application/json',
        },
      })

      expect(response.status).toBe(204)
    })

    it('should not login with old password', async () => {
      const payload = {
        grant_type: 'password',
        username: state.login.email,
        password: state.login.password,
      }

      const response = await fetch(`${state.apiURL}/v1/oauth/token`, {
        method: 'POST',
        body: JSON.stringify(payload),
        headers: {'content-type': 'application/json'},
      })

      expect(response.status).toBe(401)
    })

    it('should login with new password', async () => {
      const payload = {
        grant_type: 'password',
        username: state.login.email,
        password: 'new-password1',
      }

      const response = await fetch(`${state.apiURL}/v1/oauth/token`, {
        method: 'POST',
        body: JSON.stringify(payload),
        headers: {'content-type': 'application/json'},
      })

      expect(response.status).toBe(200)
      state.login.password = 'new-password1'
    })
  })

  require('./steps/teardown.test')(state)
})
