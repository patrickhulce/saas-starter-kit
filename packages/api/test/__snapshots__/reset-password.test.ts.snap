// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reset password reset password should reject unauthenticated password changes 1`] = `
Object {
  "message": "password link expired",
  "name": "AssertionError",
}
`;

exports[`reset password reset password should send a password reset email 1`] = `
Object {
  "content": Object {
    "from": "noreply@THE_DOMAIN",
    "html": "Hello Admin,<br><br>You're receiving this message because you requested a password reset for THE_PRODUCT_DISPLAY_NAME.<br>Please follow the link below to reset your password.<br><br><a href=\\"http://localhost/login?reset-password-key=<key>\\">http://localhost/login?reset-password-key=<key></a><br><br>Best,<br>The THE_PRODUCT_DISPLAY_NAME Team",
    "subject": "Your Password Reset - THE_PRODUCT_DISPLAY_NAME",
    "text": "Hello Admin,

You're receiving this message because you requested a password reset for THE_PRODUCT_DISPLAY_NAME.
Please follow the link below to reset your password.

http://localhost/login?reset-password-key=<key>

Best,
The THE_PRODUCT_DISPLAY_NAME Team",
  },
  "options": Object {
    "transactional": true,
  },
  "recipients": Array [
    Object {
      "address": Object {
        "email": "<uuid>@example.com",
        "name": "Admin",
      },
      "tags": Array [
        "password-reset",
      ],
    },
  ],
}
`;
